### PR TITLE
Use batch database operations instead of one by one to optimze api pulls

### LINKED ISSUE #31492: repos/pulls api take long time for large repos
### Description

I'm working with renovate which makes calls to `api/v1/repos/org/repo/pulls?state=all&sort=recentupdate&page=[page]`. At pages 8+ the requests take a long time. Looking at htop it's because the git diff command is blocking. 
This is likely because my repo had a large restructure, loading the diffs on the gitea web app is also very slow.

```
/usr/bin/git -c protocol.version=2 -c credential.helper= -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= diff --shortstat [left]...[right]
```

Is there any way to cache diffs or any config I can use to speed them up? My current workaround is to set the request timeout to 5 min in renovate

### Gitea Version

1.22.0

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

https://gist.github.com/GammaGames/b1ae6715c8a66da638ee181ed6f93da1

### Screenshots

image

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Docker compose:
```yaml
services:
  gitea:
    image: gitea/gitea:1
    container_name: gitea
    environment:
      USER_UID: 1001
      USER_GID: 1001
    restart: unless-stopped
    volumes:
      - ./gitea:/data
    ports:
      - 127.0.0.1:3001:3000
      - 127.0.0.1:2222:22
```

### Database

SQLite

### PR DESCRIPTION
**Pull request by lunny**
Resolve #31492

The response time for the Pull Requests API has improved significantly, dropping from over `2000ms` to about `350ms` on my local machine. It's about `6` times faster.

A key area for further optimization lies in batch-fetching data for `apiPullRequest.ChangedFiles, apiPullRequest.Additions, and apiPullRequest.Deletions`.

Tests `TestAPIViewPulls` does exist and new tests added.

- This PR also fixes some bugs in `GetDiff` functions.
- This PR also fixes data inconsistent in test data. For a pull request, the head branch's reference should be equal to the reference in `pull/xxx/head`.

### CODE REVIEW

File: models/issues/pull_test.go [hunk 1]
```diff
@@ -234,22 +234,6 @@ func TestPullRequest_UpdateCols(t *testing.T) {
 	unittest.CheckConsistencyFor(t, pr)
 }
 
-func TestPullRequestList_LoadAttributes(t *testing.T) {
```
Comment (lunny) [inline models/issues/pull_test.go:237]: Moved to pull_list_test.go

**Comment by vtolstov** [comment]
very useful, i have many big repos with my own code and waiting for 2-3s is very annoying

**Review by lafriks — APPROVED** [review]

**Review by Zettat123 — APPROVED** [review]

File: services/convert/pull.go [hunk 2]
```diff
@@ -259,3 +261,261 @@ func ToAPIPullRequest(ctx context.Context, pr *issues_model.PullRequest, doer *u
...
+
+	const repoDoerPermCacheKey = "repo_doer_perm_cache"
+	p, err := cache.GetWithContextCache(ctx, repoDoerPermCacheKey, fmt.Sprintf("%d_%d", baseRepo.ID, doerID),
+		func() (access_model.Permission, error) {
+			return access_model.GetUserRepoPermission(ctx, baseRepo, doer)
+		})
+	if err != nil {
+		log.Error("GetUserRepoPermission[%d]: %v", baseRepo.ID, err)
+		p.AccessMode = perm.AccessModeNone
+	}
```
Comment (wxiaoguang) [inline services/convert/pull.go]: Is this "cache" still useful?

File: services/convert/pull.go [hunk 3]
```diff
@@ -259,3 +261,261 @@ func ToAPIPullRequest(ctx context.Context, pr *issues_model.PullRequest, doer *u
 
 	return apiPullRequest
 }
+
+func ToAPIPullRequests(ctx context.Context, baseRepo *repo_model.Repository, prs issues_model.PullRequestList, doer *user_model.User) ([]*api.PullRequest, error) {
```
Comment (wxiaoguang) [inline services/convert/pull.go:265]: Is there a test to cover this function?

Comment (lunny) [reply services/convert/pull.go:265]: This has been covered by `TestAPIViewPulls` in `tests/integration/api_pull_test.go`.

Comment (wxiaoguang) [reply services/convert/pull.go:265]: There are so many fields, `api_pull_test.go` seems quite small?

File: services/convert/pull.go (hunk 2 above)
Comment (lunny) [reply services/convert/pull.go]: Removed.

File: services/convert/pull.go (hunk 3 above)
Comment (lunny) [reply services/convert/pull.go:265]: I will add more tests.

Comment (lunny) [reply services/convert/pull.go:265]: Tests added.

**Pull request merged**